opm/grid/cpgrid/Geometry.hpp

```diff
@@ -986,6 +986,425 @@ namespace Dune
+                refined_corners.resize((refined_dim[0] + 1)*(refined_dim[1] + 1)*(refined_dim[2] + 1));
+                // The nummbering starts at the botton, so k=0 (z-axis), and j=0 (y-axis), i=0 (x-axis).
+                // Then, increasing k ('going up'), followed by increasing i ('going right->'),
+                // and finally, increasing j ('going back'). This order criteria for corners
+                // 'Up [increasing k]- Right [incresing i]- Back [increasing j]'
+                // is consistant with cpgrid numbering.
+                //
+                assert(static_cast<int>(dx.size()) == patch_dim[0]);
+                assert(static_cast<int>(dy.size()) == patch_dim[1]);
+                assert(static_cast<int>(dz.size()) == patch_dim[2]);
+                const auto localCoordNumerator = []( const std::vector<double>& vec, int sumLimit, double multiplier) {
+                    double lcn = 0;
+                    assert(!vec.empty());
+                    assert(sumLimit <= static_cast<int>(vec.size()));
+                    lcn += multiplier*vec[sumLimit];
+                    for (int m = 0; m < sumLimit; ++m) {
+                        lcn += vec[m];
+                    }
+                    return lcn;
+                };
+                // E.g. localCoordNumerator( dx, 3, 0.25) =  x0 + x1 + x2 + 0.25.x3
+                //
+                const double sum_dx = std::accumulate(dx.begin(), dx.end(), double(0)); // x0 + x1 + ... + xL, if dx = {x0, x1, ..., xL}
+                const double sum_dy = std::accumulate(dy.begin(), dy.end(), double(0)); // y0 + y1 + ... + yM, if dy = {y0, y1, ..., yM}
+                const double sum_dz = std::accumulate(dz.begin(), dz.end(), double(0)); // z0 + z1 + ... + zN, if dz = {z0, z1, ..., zN}
+                std::cout << "sum_x: " << sum_dx << " sum_y: " << sum_dy << " sum_z: " << sum_dz << std::endl;
+                for (int j = 0; j < refined_dim[1] +1; ++j) {
+                    for (int i = 0; i < refined_dim[0] +1; ++i) {
+                        for (int k = 0; k < refined_dim[2] +1; ++k) {
+                            // Compute the index of each global refined corner associated with 'jik'.
+                            int refined_corner_idx =
+                                (j*(refined_dim[0]+1)*(refined_dim[2]+1)) + (i*(refined_dim[2]+1)) + k;
+                            // Compute the local refined corner of the unit/reference cube associated with 'jik'.
+                            const double local_x = localCoordNumerator(dx, i/cells_per_dim[0], double((i % cells_per_dim[0])) / cells_per_dim[0]);
+                            const double local_y = localCoordNumerator(dy, j/cells_per_dim[1], double((j % cells_per_dim[1])) / cells_per_dim[1]);
+                            const double local_z = localCoordNumerator(dz, k/cells_per_dim[2], double((k % cells_per_dim[2])) /cells_per_dim[2]);
```

Comment:
This looks really complicated. It took me a long time to understand it. That might be me, of course. I would probably have done it like this (just keeping track of the coordinates by adding to them):
```suggestion
                double local_y = 0;
                for (int j = 0; j < refined_dim[1] +1; ++j, local_y+=dy[j]) {
                    double local_x = 0;
                    for (int i = 0; i < refined_dim[0] +1; ++i, local_x+=dx[i]) {
                        double local_z = 0;
                        for (int k = 0; k < refined_dim[2] +1; ++k, local_z+=dz[k]) {
                            // Compute the index of each global refined corner associated with 'jik'.
                            int refined_corner_idx =
                                (j*(refined_dim[0]+1)*(refined_dim[2]+1)) + (i*(refined_dim[2]+1)) + k;
                            // local_x, local_y, and local z should already be correct now.
```